flake8_trio.py

```diff
@@ -40,19 +47,58 @@ class Visitor(ast.NodeVisitor):
     def __init__(self) -> None:
         super().__init__()
         self.problems: List[Error] = []
+        self.safe_yields: Set[ast.Yield] = set()
+        self._yield_is_error = False
+        self._context_manager = False
 
-    def visit_With(self, node: ast.With) -> None:
+    def visit_generic_with(self, node: Union[ast.With, ast.AsyncWith]):
         self.check_for_trio100(node)
+
+        outer = self._yield_is_error
+        if not self._context_manager and any(
+            is_trio_call(item, "open_nursery", *cancel_scope_names)
+            for item in (i.context_expr for i in node.items)
+        ):
+            self._yield_is_error = True
+
         self.generic_visit(node)
+        self._yield_is_error = outer
+
+    def visit_With(self, node: ast.With) -> None:
+        self.visit_generic_with(node)
 
     def visit_AsyncWith(self, node: ast.AsyncWith) -> None:
-        self.check_for_trio100(node)
+        self.visit_generic_with(node)
+
+    def visit_generic_FunctionDef(
+        self, node: Union[ast.FunctionDef, ast.AsyncFunctionDef]
+    ):
+        outer = self._context_manager
+        if any(
+            isinstance(d, ast.Name)
+            and d.id in ("contextmanager", "asynccontextmanager")
+            for d in node.decorator_list
+        ):
+            self._context_manager = True
+        self.generic_visit(node)
```

Comment:
We *also* want to un-set `self._yield_is_error` here; it's OK if you *define* a generator function inside a nursery or cancel scope - just not yield from it.

- If you're inside a contextmanager-decorated function, yield is OK
- Otherwise, if you're in a cancel scope or nursery, yield is an error
- *Unless* you're inside a function being defined in that scope, then it's OK again

Reply:
isn't that the current behaviour? The default value is false, and it's only set to true within the cancel scope (if it's not w/in a context manager)

Reply:
```python
async def f():
    async with trio.open_nursery():
        yield 1  # is and should be an error

        def g():
            yield 2  # is but should not be an error
```

Reply:
yep, you're right. Fixed~